Common/cpp/LayoutAnimations/LayoutAnimationsProxy.cpp

```diff
@@ -11,35 +11,100 @@ namespace reanimated {
 const long long idOffset = 1e9;
 
 LayoutAnimationsProxy::LayoutAnimationsProxy(
-    std::function<void(int, jsi::Object newProps)> _notifyAboutProgress,
-    std::function<void(int, bool)> _notifyAboutEnd)
-    : notifyAboutProgress(std::move(_notifyAboutProgress)),
-      notifyAboutEnd(std::move(_notifyAboutEnd)) {}
+    std::function<void(int, jsi::Object newProps)> progressHandler,
+    std::function<void(int, bool, std::string)> endHandler)
+    : progressHandler(std::move(progressHandler)),
+      endHandler(std::move(endHandler)) {}
 
 void LayoutAnimationsProxy::startObserving(
     int tag,
     std::shared_ptr<MutableValue> sv,
     jsi::Runtime &rt) {
   observedValues[tag] = sv;
+  this->progressHandler(tag, sv->value->toJSValue(rt).asObject(rt));
   sv->addListener(tag + idOffset, [sv, tag, this, &rt]() {
-    std::shared_ptr<FrozenObject> newValue =
-        ValueWrapper::asFrozenObject(sv->value->valueContainer);
-    this->notifyAboutProgress(tag, newValue->shallowClone(rt));
+    this->progressHandler(tag, sv->value->toJSValue(rt).asObject(rt));
   });
 }
 
-void LayoutAnimationsProxy::stopObserving(int tag, bool finished) {
+void LayoutAnimationsProxy::stopObserving(
+    int tag,
+    bool finished,
+    std::string type) {
   if (observedValues.count(tag) == 0) {
     return;
   }
   std::shared_ptr<MutableValue> sv = observedValues[tag];
   sv->removeListener(tag + idOffset);
   observedValues.erase(tag);
-  this->notifyAboutEnd(tag, !finished);
+  this->endHandler(tag, !finished, type);
 }
 
-void LayoutAnimationsProxy::notifyAboutCancellation(int tag) {
-  this->notifyAboutEnd(tag, false);
+void LayoutAnimationsProxy::configureAnimation(
+    int tag,
+    const std::string &type,
+    std::shared_ptr<ShareableValue> config,
+    std::shared_ptr<ShareableValue> viewSharedValue) {
+  auto lock = std::unique_lock<std::mutex>(animationsLock);
+  if (type == "entering") {
+    enteringAnimations[tag] = config;
+  } else if (type == "exiting") {
+    exitingAnimations[tag] = config;
+  } else if (type == "layout") {
+    layoutAnimations[tag] = config;
+  }
+  viewSharedValues[tag] = viewSharedValue;
+}
+
+bool LayoutAnimationsProxy::hasLayoutAnimation(
+    int tag,
+    const std::string &type) {
+  auto lock = std::unique_lock<std::mutex>(animationsLock);
+  if (type == "entering") {
```

Comment:
Do we really need `type` as string? Could it be an enum type?

Reply:
Let's leave it for later, since it requires changing a lot.